src/System.Memory/ref/System.Memory.cs

```diff
@@ -178,7 +178,8 @@ public unsafe struct MemoryHandle : IDisposable
     {
         public MemoryHandle(IRetainable owner, void* pinnedPointer = null,  System.Runtime.InteropServices.GCHandle handle = default(System.Runtime.InteropServices.GCHandle))  { throw null; }
         public void* PinnedPointer { get { throw null; } }
-        public void Dispose()  { throw null; }
+        internal void AddOffset(int offset) { throw null; }
```

Comment:
Why does an internal method need to be in the ref?

Reply:
I suppose it doesn't.

I will remove internal methods in the ref in other places as well.
- https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L131
- https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L203
- etc.

Are internal constructors treated uniquely? I see a lot of internal constructors in the S.Runtime ref.
https://github.com/dotnet/corefx/blob/master/src/System.Runtime/ref/System.Runtime.cs#L209

There are some other places that have internal APIs in the ref. Should they be there?
https://github.com/dotnet/corefx/blob/master/src/System.Threading/ref/System.Threading.cs#L139

Reply:
The only reason they'd need to be there is if some other library has InternalsVisibleTo access and is accessing them.  We've tried to eliminate all such usage, though maybe there are still a few stragglers.

Reply:
Wouldn't the InternalsVisibleTo apply only to the source projects? Why would that require internals to be listed in the reference assembly?

I mainly see InternalsVisibleTo for the test project related to the library itself: https://github.com/dotnet/corefx/search?p=1&q=InternalsVisibleTo&type=&utf8=%E2%9C%93

But in some cases (like the one I linked), I don't see any library that is accessing this internal API (other than the [test](https://github.com/dotnet/corefx/blob/master/src/System.Threading/tests/HostExecutionContextTests.cs#L51)):
https://github.com/dotnet/corefx/blob/master/src/System.Threading/ref/System.Threading.cs#L139
```C#
    public partial class HostExecutionContext : System.IDisposable
    {
        public HostExecutionContext() { }
        public HostExecutionContext(object state) { }
        protected internal object State { get { throw null; } set { } }
        public virtual System.Threading.HostExecutionContext CreateCopy() { throw null; }
        public void Dispose() { }
        public virtual void Dispose(bool disposing) { }
    }
```
However, if I remove it from the ref, I get APICompat error:
`error : MembersMustExist : Member 'System.Threading.HostExecutionContext.State.get()' does not exist in the implementation but it does exist in the contract.`

Reply:
> Are internal constructors treated uniquely

Yes. The C# compiler generates a public constructor if you do not have any constructor. If you do not want to have a public constructor, you need to have internal one - it is the purpose of the internal constructors in S.Runtime ref.

Reply:
> However, if I remove it from the ref,

What did you remove, just the internal keyword, or the whole property?  "protected internal" means "protected or internal", not "protected and internal", so a "protected internal" property is visible outside of the assembly as protected.

Reply:
> If you do not want to have a public constructor, you need to have internal one - it is the purpose of the internal constructors in S.Runtime ref.

Got it. That makes sense.

> What did you remove, just the internal keyword, or the whole property? "protected internal" means "protected or internal", not "protected and internal", so a "protected internal" property is visible outside of the assembly as protected.

I had removed the whole property (thinking it was "and"). Thanks for the clarification.